src/game/Entities/Unit.cpp

```diff
@@ -6054,44 +6058,78 @@ void Unit::ModifyAuraState(AuraState flag, bool apply)
     }
 }
 
-Unit* Unit::GetOwner() const
+ObjectGuid const& Unit::GetMasterGuid() const
 {
-    if (ObjectGuid ownerid = GetOwnerGuid())
-        return ObjectAccessor::GetUnit(*this, ownerid);
-    return nullptr;
+    ObjectGuid const& guid = GetCharmerGuid();
+    return (guid ? guid : GetOwnerGuid());
 }
 
-Unit* Unit::GetCharmer() const
+Unit* Unit::GetOwner(bool recursive /*= false*/) const
 {
-    if (ObjectGuid charmerid = GetCharmerGuid())
-        return ObjectAccessor::GetUnit(*this, charmerid);
-    return nullptr;
+    // Default, creator field as owner is present in everything: totems, pets, guardians, etc
+    Unit* owner = GetCreator();
+    // Query owner recursively (ascending)
+    if (recursive)
+    {
+        while (Unit* grandowner = (owner ? owner->GetOwner() : nullptr))
+            owner = grandowner;
+    }
+    return owner;
 }
 
-bool Unit::IsCharmerOrOwnerPlayerOrPlayerItself() const
+Unit* Unit::GetMaster() const
 {
-    if (GetTypeId() == TYPEID_PLAYER)
-        return true;
+    Unit* charmer = GetCharmer();
+    return (charmer ? charmer : GetOwner());
+}
 
-    return GetCharmerOrOwnerGuid().IsPlayer();
+Unit const* Unit::GetBeneficiary() const
+{
+    Unit const* master = GetMaster();
+    return (master ? master : this);
 }
 
-Player* Unit::GetCharmerOrOwnerPlayerOrPlayerItself()
+Unit* Unit::GetBeneficiary()
 {
-    ObjectGuid guid = GetCharmerOrOwnerGuid();
-    if (guid.IsPlayer())
-        return ObjectAccessor::FindPlayer(guid);
+    Unit* master = GetMaster();
+    return (master ? master : this);
+}
 
-    return GetTypeId() == TYPEID_PLAYER ? (Player*)this : nullptr;
+Player const* Unit::GetBeneficiaryPlayer() const
+{
+    Unit const* beneficiary = GetBeneficiary();
+    if (beneficiary)
+        return (beneficiary->GetTypeId() == TYPEID_PLAYER ? static_cast<Player const*>(beneficiary) : nullptr);
+    return (GetTypeId() == TYPEID_PLAYER ? static_cast<Player const*>(this) : nullptr);
 }
 
-Player const* Unit::GetCharmerOrOwnerPlayerOrPlayerItself() const
+Player* Unit::GetBeneficiaryPlayer()
 {
-    ObjectGuid guid = GetCharmerOrOwnerGuid();
-    if (guid.IsPlayer())
-        return ObjectAccessor::FindPlayer(guid);
+    Unit* beneficiary = GetBeneficiary();
+    if (beneficiary)
+        return (beneficiary->GetTypeId() == TYPEID_PLAYER ? static_cast<Player*>(beneficiary) : nullptr);
+    return (GetTypeId() == TYPEID_PLAYER ? static_cast<Player*>(this) : nullptr);
+}
 
-    return GetTypeId() == TYPEID_PLAYER ? (Player const*)this : nullptr;
+Unit* Unit::GetSummoner() const
+{
+    if (ObjectGuid guid = GetSummonerGuid())
```

Comment:
ObjectGuid const& guid (even if GetUnit argument will need to be changed too if we would make it more sence)